Implemented ScrollItemPattern for ComboBoxItemAccessibleObject

vladimir-krestov opened the pull request:
Fixes #4335
Fixes #4323

## Proposed changes

- Implement `ScrollIntoView` method of `ScrollItemPattern`
- Fix getting of `Bounds`

<!-- We are in TELL-MODE the following section must be completed -->

## Customer Impact

- Accessibility tools or users can scroll a combo box drop-down list to some item
- A user can see a correct item Bounds position when combobox scrolling

## Regression? 

- Yes (from .NET Framework 4.7.2)

## Risk

- Minimal


## Screenshots <!-- Remove this section if PR does not change UI -->

### Before
- ComboBox items don't support `ScrollItemPattern` - do nothing when `ScrollIntoView` calling
![image](https://user-images.githubusercontent.com/49272759/101989990-070d6e80-3cb5-11eb-8fa0-073a0a844a43.png)



### After
- ComboBox items support `ScrollItemPattern` - scroll to some item when `ScrollIntoView` calling. 
For example, in the gif below I call `ScroolIntoView` for the **9th** item, and the `ComboBox` scrolls to the **9th** item on top or bottom.
![sU2z22IY0t](https://user-images.githubusercontent.com/49272759/101990241-e47c5500-3cb6-11eb-8d14-7d7804a9fa5b.gif)



## Test methodology <!-- How did you ensure quality? -->

- Unit tests
- Manual testing
- CTI

## Accessibility testing  <!-- Remove this section if PR does not change UI -->
- Using Inspect
<!--
     Microsoft prioritizes making our products accessible. 
     WinForms has a key role in allowing developers to create accessible apps. 
     
     When submitting a change which impacts UI in any way, including adding new UI or
     modifying existing controls the developer needs to run the Accessibility Insights
     tool (https://accessibilityinsights.io/) and verify that there are no changes or
     regressions. 
     
     The developer should run the Fast Pass over the impacted control(s) and provide
     a snapshot of the passing results along with before/after snapshots of the UI.
     More info: (https://accessibilityinsights.io/docs/en/web/getstarted/fastpass)
  -->


 

## Test environment(s) <!-- Remove any that don't apply -->

- Microsoft Windows [Version 10.0.19042.685]
- .NET v. 6.0.100-alpha.1.20554.10

<!-- Mention language, UI scaling, or anything else that might be relevant -->


###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/4350)

vladimir-krestov commented:
Just a draft. Needs more testing and improvements.

Tanya-Solyanik commented on src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObject.cs:
```diff
@@ -246,6 +249,57 @@ public override AccessibleStates State
+
+                if (_owningComboBox.DropDownStyle == ComboBoxStyle.Simple)
+                {
+                    visibleItemsCount = (_owningComboBox.Size.Height - 26) / itemHeight; // 26 is an edit control hight (23) + 3px of borders
```
Would CB_GETMINVISIBLE  work?

vladimir-krestov replied:
@Tanya-Solyanik 
No. It always returns 30.
```cs
int visibleCount = User32.SendMessageW(_owningComboBox, (User32.WM)User32.CB.GETMINVISIBLE, (IntPtr)currentIndex).ToInt32(); // = 30
```
I spent much time to find `CB_GETMINVISIBLE` value, because we don't have it in the project and the documentation.
`SETMINVISIBLE = 0x1701,`
`GETMINVISIBLE = 0x1702`

Actually, our `ComboBox` works incorrectly.
We have the following ComboBox properties:
- [IntegralHeight](https://docs.microsoft.com/dotnet/api/system.windows.forms.combobox.integralheight)
- [DropDownHeight](https://docs.microsoft.com/dotnet/api/system.windows.forms.combobox.dropdownheight)
- [MaxDropDownItems](https://docs.microsoft.com/dotnet/api/system.windows.forms.combobox.maxdropdownitems)

They don't reflect on each other.
- `MaxDropDownItems` has **8** as the default value. We can see 8 items in a drop-down if to set `IntegralHeight` to **`false`** only and if `DropDownHeight` is not set. If `IntegralHeight` is true or `DropDownHeight` is set then `MaxDropDownItems` doesn't reflect a real count of visible items. For example:
```cs
this.comboBox3.IntegralHeight = false;
this.comboBox3.DropDownHeight = 131; // 8.5 items are visible. One item is 15px height
this.comboBox3.MaxDropDownItems = 5; // All the same we see 8.5 items
```
- `MaxDropDownItems` and `DropDownHeight` don't affect a drop-down height of a `ComboBox` with the `Simple` drop-down style. And these properties will store not real values.
- `MaxDropDownItems` and `DropDownHeight` do not matter if `IntegralHeight` is **`true`** (default value) and they store not real visible values.
- If `MaxDropDownItems` is set then `DropDownHeight` doesn't reflect a real value. For example:
```cs
this.comboBox3.IntegralHeight = false;
this.comboBox3.MaxDropDownItems = 5; // The real drop-down height is about 75 px (5*15px)
//this.comboBox3.DropDownHeight = 131; // Is not set, but if we try to get this value we will get 106 px
```

I expect that these 3 properties should reflect real visible values when some of them are set. I think, we can discuss it and fix it as another Issue. We should use `CB_SETMINVISIBLE` to set correct values for a native control.

vladimir-krestov replied:
At this moment, I'm working on getting a real value of a drop-down height from a native window. It will allow us to a get real count of visible items.

vladimir-krestov replied:
Have made it. Please review.

RussKie replied:
> We have the following ComboBox properties...(snip)
> They don't reflect on each other.

The docs state that setting `DropDownHeight` will reset `IntegralHeight` to `false`

vladimir-krestov replied:
Yes, you are right, that is a little worked point. 
But if you set `IntegralHeight` as true after that, then the visible items count will be **30** and `DropDownHeight` will not reflect real value.

vladimir-krestov commented on src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObject.cs:
```diff
@@ -157,6 +159,16 @@ internal override int GetChildId()
+                ComboBoxChildNativeWindow dropDownWindow = _owningComboBox.DropDownStyle == ComboBoxStyle.Simple
+                    ? _owningComboBox._childListBox : _owningComboBox._childDropDown;
+
+                return (dropDownWindow.Size.Height + 2) / itemHeight; // 2px of borders
```
It works well in the high DPI mode.
Sometimes dropdown window size lacks a couple of pixels and we add it here to get the correct visible items count.

vladimir-krestov commented:
Tested in HDPI mode and without enabled visual styles. The algorithm works well.

RussKie commented:
> Tested in HDPI mode, in High Contrast mode and without enabled visual styles.

Got screenshots? :)

RussKie reviewed (commented):
LGTM

vladimir-krestov commented:
> > Tested in HDPI mode, in High Contrast mode and without enabled visual styles.
> 
> Got screenshots? :)

- DPI 150%
![1](https://user-images.githubusercontent.com/49272759/102209189-7667a600-3ee1-11eb-83bc-d932c1f96e2d.gif)

- HighContrast
![2](https://user-images.githubusercontent.com/49272759/102209228-87181c00-3ee1-11eb-99a2-ab6e12fccc08.gif)

- Without visual styles
![3](https://user-images.githubusercontent.com/49272759/102209260-9008ed80-3ee1-11eb-9977-8317467fd685.gif)

Tanya-Solyanik commented on src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObject.cs:
```diff
@@ -171,6 +183,7 @@ internal override bool IsPatternSupported(UiaCore.UIA patternId)
             {
                 if (patternId == UiaCore.UIA.LegacyIAccessiblePatternId ||
                     patternId == UiaCore.UIA.InvokePatternId ||
+                    patternId == UiaCore.UIA.ScrollItemPatternId ||
```
Is this pattern always available in 4.7.2, or only when the scroll bar is displayed?

vladimir-krestov replied:
A good question, will check it in 4.7.2.

vladimir-krestov replied:
Yes

![image](https://user-images.githubusercontent.com/49272759/102521357-e4f56100-40a5-11eb-9573-126da6c3ff3f.png)

Tanya-Solyanik commented on src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObject.cs:
```diff
@@ -246,6 +259,40 @@ public override AccessibleStates State
+                    return;
+                }
+
+                int currentIndex = _owningComboBox.Items.IndexOf(_owningItem);
```
Please set the top item instead of this calculation.

vladimir-krestov replied:
I oriented to our `ListBox` and thought that the current behavior is better. But I checked **4.7.2** and a native `ComboBox` (that the same) and `ListBox`. They work as you said. They just set a top index. I'll make it as there to be consistent.

vladimir-krestov replied:
Done and tested

Tanya-Solyanik replied:
Sounds like we should change ListBox to "set top" as well, please open a new issue.

vladimir-krestov replied:
https://github.com/dotnet/winforms/issues/4392

Tanya-Solyanik commented on src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObject.cs:
```diff
@@ -235,6 +238,17 @@ public override AccessibleStates State
 
+            internal override void ScrollIntoView()
+            {
+                if (!_owningComboBox.IsHandleCreated || _owningComboBox.Items.Count == 0)
```
What happens if count is zero? Do we need this test in NETFX?

Tanya-Solyanik replied:
This file assumes that the _owningItem is in the list of items of this combobox, I don't see the reason to test for an empty list here.

vladimir-krestov replied:
>  I.e. what happens if CB_SETTOPINDEX attempts to set index to -1?

`SendMessageW` returns "-1" as the result and do nothing.

> I don't see the reason to test for an empty list here.

I agree, I'll remove this.

Tanya-Solyanik reviewed (changes requested):
Please wait with this change, I need to verify if we should sent set top message only after we validated that the item is not shown already.

vladimir-krestov commented:
> Please wait with this change, I need to verify if we should sent set top message only after we validated that the item is not shown already.

I think, you are right and we need to check if an item is visible and don't send a message, because a native ListBox works like that.

Tanya-Solyanik commented on src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObject.cs:
```diff
@@ -235,6 +238,17 @@ public override AccessibleStates State
+                }
+
+                int currentIndex = GetCurrentIndex();
+                User32.SendMessageW(_owningComboBox, (User32.WM)User32.CB.SETTOPINDEX, (IntPtr)currentIndex);
```
MSAAProxy was scrolling only if the item had not been already visible, we should match that behavior.

vladimir-krestov replied:
Testing different cases...

vladimir-krestov replied:
Thank you! Added `IntersectsWith` above

Tanya-Solyanik reviewed (commented):
Looks good, I have a minor comment about checking the zero return value from MapWindowPoints. 
What is your plan for unit tests?

For the bounds change, please also add a manual test to `WinFormsControls` app

```
  internal class MyCombo : ComboBox
    {
        public MyCombo() : base()
        {
            DrawMode = DrawMode.OwnerDrawVariable;
            DropDownStyle = ComboBoxStyle.DropDownList;
            MaxDropDownItems = 7;
            ItemHeight = 13;
            IntegralHeight = false;

            for (int i = 0; i < 10; i++)
            {
                Items.Add($"Item {i}");
            }

            SelectedIndex = 0;
        }

        protected override void OnMeasureItem(MeasureItemEventArgs e)
        {
            e.ItemHeight = 13 + e.Index * 5;
        }

        protected override void OnDrawItem(DrawItemEventArgs e)
        {
            e.DrawBackground();

            using var brush = new SolidBrush(e.ForeColor);

            e.Graphics.DrawString(
                Items[e.Index].ToString(),
                e.Font,
                brush,
                e.Bounds);

            e.DrawFocusRectangle();
        }
    }
```

RussKie commented on src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObject.cs:
```diff
@@ -42,23 +42,18 @@ public override Rectangle Bounds
 
-                        // Using the first visible index, we make an index shift, which helps to draw a rectangle with the correct position
-                        currentIndex -= firstVisibleIndex;
+                    if (User32.SendMessageW(listHandle, (User32.WM)User32.LB.GETITEMRECT,
+                        (IntPtr)currentIndex, ref itemRect).ToInt32() == User32.LB_ERR ||
+                        User32.MapWindowPoints(listHandle, IntPtr.Zero, ref itemRect, 2) == 0)
```
This is unreadable, there are multiple method calls and code paths. Break it up.

RussKie commented on src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObject.cs:
```diff
@@ -42,23 +42,18 @@ public override Rectangle Bounds
 
-                        // Using the first visible index, we make an index shift, which helps to draw a rectangle with the correct position
-                        currentIndex -= firstVisibleIndex;
+                    if (User32.SendMessageW(listHandle, (User32.WM)User32.LB.GETITEMRECT,
```
A single PR is fine.

RussKie replied:
A single PR with distinct commits is fine.

vladimir-krestov replied:
Will do 2 commits in this PR

vladimir-krestov replied:
Added 3 commits (+2 temporary for review).
- Add ScrollItemPattern + tests
- Fixed Bounds + tests
- Add a UI test case to test ComboBoxes with scrollbars and different heights items

RussKie commented on src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObject.cs:
```diff
@@ -42,23 +42,18 @@ public override Rectangle Bounds
-                        // Using the first visible index, we make an index shift, which helps to draw a rectangle with the correct position
-                        currentIndex -= firstVisibleIndex;
+                    if (User32.SendMessageW(listHandle, (User32.WM)User32.LB.GETITEMRECT,
+                        (IntPtr)currentIndex, ref itemRect).ToInt32() == User32.LB_ERR ||
```
❗  `.ToInt32()` will overflow on x86!

vladimir-krestov replied:
Should I use `(int)(long)` cast here?

vladimir-krestov commented:
I didn't add a unit test that checks `ScrollIntoView` code path when the owning combobox doesn't have a Handle because `ComboBoxItemAccessibleObject` can't exist in this case.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/ComboBoxesWithScrollBars.cs:
```diff
@@ -0,0 +1,193 @@
+        {
+            ComboBox control = (ComboBox)sender;
+
+            if (e.Index < 0)
```
I don't think it's possible that the index in the event args is off. But if you want to check bounds, please check the upper bound as well.

vladimir-krestov replied:
I got "-1" Index a couple of times in my previous test app (not here), so I added this check. I would check the bounds. Thanks, I added upper bound as well

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,386 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+        public void ComboBoxItemAccessibleObject_IsPatternSupported_ReturnsTrue_ForScrollItemPattern()
+        {
+            using ComboBox comboBox = new ComboBox();
+            comboBox.Items.AddRange(new string[] { "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10" });
```
Please run Ctrl-K + Ctrl-D on this file to apply the default formatting. I think there should not be a space between the brace and the first or last item.

Also I think `new string[]` can be simplified to `new[]`.

vladimir-krestov replied:
1) The default formatting is applying when I'm saving files
2) `Ctrl-K` + `Ctrl-D` is my reflex 😊 (doing it after almost every change)
     So, here is the result of the default VS formatting.
     
Thanks, I simplified to `new[]`

vladimir-krestov replied:
![DkJ8950mEt](https://user-images.githubusercontent.com/49272759/103344128-b3fe2e80-4a9e-11eb-8b3d-aba30a776993.gif)

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,386 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+                ComboBoxItemAccessibleObject itemAccessibleObject = itemsCollection.GetComboBoxItemAccessibleObject(itemEntry);
+
+                // Check that all items support ScrollItemPattern
+                result &= itemAccessibleObject.IsPatternSupported(UiaCore.UIA.ScrollItemPatternId);
```
I think that this operator short-circuits when `result` is `false`, so once you hit an item that does not support this pattern, you are no longer testing the rest of the items anyway. Also you don't report the failed items. So I don't see the reason to continue once the bad item is encountered. 
I think   
`Assert.True(itemAccessibleObject.IsPatternSupported(UiaCore.UIA.ScrollItemPatternId));`
is cleaner and will show the first failed item.

vladimir-krestov replied:
Thank you, I agree, your approach is more readable. Have reworked it

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,386 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+                ComboBoxItemAccessibleObject itemAccessibleObject = itemsCollection.GetComboBoxItemAccessibleObject(itemEntry);
+
+                // Check that all items support ScrollItemPattern
+                result &= (bool)itemAccessibleObject.GetPropertyValue(UiaCore.UIA.IsScrollItemPatternAvailablePropertyId);
```
Assert.True(itemAccessibleObject.GetPropertyValue(UiaCore.UIA.IsScrollItemPatternAvailablePropertyId));

Or consider factoring out the foreach loop into a helper function and share it with the above test.

vladimir-krestov replied:
I think put `Assert` inside the cycle is more readable.
I think we don't need to create a new function for the sake of 2 places and 3 lines. It will complicate the tests

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,386 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+            Assert.True(result);
+        }
+
+        public static IEnumerable<object[]> ComboBoxItemAccessibleObject_ScrollIntoView_DoNothing_IfControlIsNotEnabled_TestDate()
```
TestDate-> TestData

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,386 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+        [MemberData(nameof(ComboBoxItemAccessibleObject_ScrollIntoView_DoNothing_IfControlIsNotEnabled_TestDate))]
+        public void ComboBoxItemAccessibleObject_ScrollIntoView_DoNothing_IfControlIsNotEnabled(ComboBoxStyle comboBoxStyle)
+        {
+            using ComboBox comboBox = new ComboBox();
```
FYI - in this context, i.e. explicit type in this declaration,  `new ComboBox()` could be simplified to `new()`. This applies to other changes in this file as well.

vladimir-krestov replied:
Yes, it is possible, but I think this change won't be consistent with the rest unit tests because many test files have explicit definitions of objects.
Actually, it may be changed later as a code refactoring. But not now.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,386 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+            Entry itemEntry = comboBox.Items.InnerList[itemIndex];
+            ComboBoxItemAccessibleObject itemAccessibleObject = itemsCollection.GetComboBoxItemAccessibleObject(itemEntry);
+            itemAccessibleObject.ScrollIntoView();
+            int topIndex = unchecked((int)(long)User32.SendMessageW(comboBox, (User32.WM)User32.CB.GETTOPINDEX));
```
Consider naming this variable `actual` or `actualTopIndex`

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,386 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+            ComboBoxItemAccessibleObjectCollection itemsCollection = comboBoxAccessibleObject.ItemAccessibleObjects;
+            Entry itemEntry = comboBox.Items.InnerList[itemIndex];
+            ComboBoxItemAccessibleObject itemAccessibleObject = itemsCollection.GetComboBoxItemAccessibleObject(itemEntry);
+            itemAccessibleObject.ScrollIntoView();
```
Consider formatting tests so that code that performs the main action is surrounded by empty lines. I.e. add blank line above and below the `ScrollIntoView` call.

vladimir-krestov replied:
Make sense

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,386 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+            {
+                int itemsCount = 21;
+
+                // In the tested comboboxes only 7 items are visible. It needs to scroll to see the rest
```
FYI - this assumes certain font height and HDPI scaling on the test machine. It might be more reliable to actually calculate if drop down rectangle intersects with the item rectangle. Another advantage of 'intersectsWith` approach is that you'd have a single test for the "do nothing" and "scroll" tests, instead of copy-pasting the test code.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,386 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+                int itemsCount = 21;
+                int visibleItemsCount = 7; // 7 items are visible for the tested combobox size
+
+                // Get an index of the first item that is visible if to scroll a combobox to the bottom
```
if to scroll a combobox to the bottom -> if dropdown is scrolled to the bottom?

Tanya-Solyanik reviewed (commented):
Nice manual test!
In the unit tests, please consider introducing helper functions to avoid copy/pasted initialization and verification code where possible. 
Also the magic number 7  might stop working at some point on some machines. Consider re-doing tests to actually calculate the if the item is visible in the test body.

RussKie commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/ComboBoxesWithScrollBars.cs:
```diff
@@ -0,0 +1,193 @@
+            }
+        }
+
+        private void ComboBoxesWithScrollBars_Load(object sender, EventArgs e)
```
Don't create an event handler, override `OnLoad` instead

vladimir-krestov replied:
I can, but are there any specific reasons?

RussKie replied:
Yes, event handlers cause extra allocations. Overriding a method doesn't.

RussKie commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/ComboBoxesWithScrollBars.cs:
```diff
@@ -0,0 +1,193 @@
+                "\nIt is impossible to return DropDownList style for the combobox");
+
+            // Add event handlers
+            changeDDH_UpDown1.ValueChanged += changeDDH_UpDown1_ValueChanged;
+            changeCBHeight_UpDown2.ValueChanged += changeCBHeight_UpDown2_ValueChanged;
+            changeDDH_UpDown3.ValueChanged += changeDDH_UpDown3_ValueChanged;
+            maxDropDownItemsUpDown1.ValueChanged += maxDropDownItemsUpDown1_ValueChanged;
+            maxDropDownItemsUpDown2.ValueChanged += maxDropDownItemsUpDown2_ValueChanged;
+            maxDropDownItemsUpDown3.ValueChanged += maxDropDownItemsUpDown3_ValueChanged;
+            integralHeightCheckBox1.CheckedChanged += integralHeightCheckBox1_CheckedChanged;
+            integralHeightCheckBox2.CheckedChanged += integralHeightCheckBox2_CheckedChanged;
+            integralHeightCheckBox3.CheckedChanged += integralHeightCheckBox3_CheckedChanged;
+            useDifferentHeightsCheckBox1.CheckedChanged += useDifferentHeightsCheckBox1_CheckedChanged;
+            useDifferentHeightsCheckBox2.CheckedChanged += useDifferentHeightsCheckBox2_CheckedChanged;
+            useDifferentHeightsCheckBox3.CheckedChanged += useDifferentHeightsCheckBox3_CheckedChanged;
```
Any specific reason not to add them via the designer?

vladimir-krestov replied:
1) If we add these event handlers in the designer, they will be invoked when we set the values above. We don't need these redundant invocations, I guess. We should invoke the handler when a user changes the values.
2) It's more readable, in my opinion. And it doesn't depend on the designer changes.

RussKie replied:
It is worth adding a comment explaining just that.

RussKie commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/ComboBoxesWithScrollBars.cs:
```diff
@@ -0,0 +1,193 @@
+            comboBox3.DrawItem += comboBox_DrawItem;
+
+            // Set additional info
+            toolTip1.SetToolTip(maxDropDownItemsUpDown2, "It doesn't affect the combobox with the Simple dropdown style");
+            toolTip1.SetToolTip(changeDDH_UpDown1, "Sets IntegralHeight to false. MaxDropDownItems won't be available after it setting");
+            toolTip1.SetToolTip(changeDDH_UpDown3, "Sets IntegralHeight to false. MaxDropDownItems won't be available after it setting");
+            toolTip1.SetToolTip(useDifferentHeightsCheckBox3, "Changes DropDownList dropdown style to DropDown in view." +
+                "\nIt is impossible to return DropDownList style for the combobox");
```
ditto

vladimir-krestov replied:
No, thanks. Moved it to the designer

vladimir-krestov commented:
I have changed the UI test app implementation in accordance with the `Designer`. Steps:
1) Created a new .NET 6.0 app
2) Copied my test code and pasted it to the new .NET 6.0 app
3) Opened the `Designer` in this app
4) Made a little change to make the `Designer` change the implementation as it wants
5) Saved the `Designer` changes and moved them to my test implementation

Now you can see the implementation created by the `Designer` in the `ComboBoxesWithScrollBars.Designer.cs` file.
All strange changes are its work.

RussKie commented on src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/ComboBoxesWithScrollBars.cs:
```diff
@@ -0,0 +1,188 @@
+
+        private void changeDDH_UpDown1_ValueChanged(object sender, EventArgs e)
+        {
+            comboBox1.DropDownHeight = (int)((NumericUpDown)sender).Value;
```
```suggestion
            comboBox1.DropDownHeight = changeDDH_UpDown1.Value;
```

In fact, since there's only one instruction here, and the handler is exclusive to `changeDDH_UpDown1`, this all can be inlined, thus drastically reducing the code:
```cs
changeDDH_UpDown1.ValueChanged += (s, e) => comboBox1.DropDownHeight = changeDDH_UpDown1.Value;
```

ditto below

vladimir-krestov commented:
Waiting for results from CTI

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,312 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+
+            protected override void OnDrawItem(DrawItemEventArgs e)
+            {
+                if (e.Index < 0)
```
Consider testing the upper bound for sanity

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,312 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+
+        [WinFormsTheory]
+        [MemberData(nameof(ComboBoxItemAccessibleObject_Bounds_ReturnsCorrect_ForVisibleItems_IfComboBoxIsScrollable_TestData))]
+        public void ComboBoxItemAccessibleObject_Bounds_ReturnsCorrect_ForVisibleItems_IfComboBoxIsScrollable(ComboBoxStyle comboBoxStyle, int itemIndex, Point expectedPosition)
```
Why do you restrict this and the next test to visible items only? Bounds rectangle should be accurate even if the item is not visible. These tests would be applicable to OffScreen and Visible properties.

vladimir-krestov replied:
You are right. I wanted to prevent failing these tests after fixing Bounds for invisible items. But it will be another story and it may be fixed in another way.
Thank you, I have fixed it, and now the tests work for all items.

Tanya-Solyanik replied:
I don't think we should return an empty rectangle to invisible items. This is not the behavior we see in other frameworks. Users(or automation tests) who work with applications that use both winforms and win32 controls would not expect that.

vladimir-krestov commented:
Testers approved the fix ✔️

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,311 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+
+        [WinFormsTheory]
+        [MemberData(nameof(ComboBoxItemAccessibleObject_ScrollIntoView_TestData))]
+        public void ComboBoxItemAccessibleObject_ScrollIntoView_DoExpected(ComboBoxStyle comboBoxStyle, bool scrollingDown, int itemIndex, int itemsCount)
```
nit: `DoExpected` is not a descriptive name. All tests are supposed to do what is expected :) I would say `ComboBoxItemAccessibleObject_ScrollIntoView_EnsureVisible`

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -135,5 +139,311 @@ public void ComboBoxItemAccessibleObject_SeveralSameItems_FragmentNavigate_Previ
+            ComboBoxItemAccessibleObject itemAccessibleObject = itemsCollection.GetComboBoxItemAccessibleObject(itemEntry);
+            Rectangle actual = itemAccessibleObject.Bounds;
+            Rectangle dropdownRect = comboBox.ChildListAccessibleObject.Bounds;
+            int itemWidth = comboBoxStyle == ComboBoxStyle.Simple ? 79 : 81;
```
Could you please add a comment about where these items are coming from?

vladimir-krestov replied:
Thanks, I added little comments.
Actually, we get items rectangles from Windows.
It returns to us different widths values depending on a combobox drop-down style

Tanya-Solyanik replied:
I meant `where these widths are coming from?`

Are you sure it is based on the style and not on the DPI or font? I think it's sufficient to test that all widths are the same rather than are equal to a specific number.

RussKie replied:
All tests are currently run at 100% and with styles enabled. We currently have no ability to run automated tests with styles disabled.
In tests we currently have hardcoded values for styles on/off scenarios, so this is inline with the rest of tests. Not having these values hardcoded is probably not a bad idea, but this should be done separately and across all tests at the same time.
